# Businnes analysis

## Data overwiev

### Download all libraries

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
from scipy import stats as st
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import ipywidgets as widgets
from IPython.display import display, clear_output

print("All libraries are downloaded.")

All libraries are downloaded.


### First look of datasets

In [2]:
cost = pd.read_csv('costs_us.csv')
orders = pd.read_csv('orders_log_us.csv')
visit = pd.read_csv('visits_log_us.csv')

print('All datasets are downloaded.')

All datasets are downloaded.


In [3]:
print('----Costs dataset----')
display(cost.sample(5))
print('-------------------------------------------------------------------')
print('----Orders dataset----')
display(orders.sample(5))
print('-------------------------------------------------------------------')
print('----Visits dataset----')
display(visit.sample(5))

----Costs dataset----


,source_id,dt,costs
1758,5,2018-04-04,108.56
157,1,2017-11-05,60.50
1895,9,2017-08-19,7.68
213,1,2017-12-31,39.66
1609,5,2017-11-05,162.07


-------------------------------------------------------------------
----Orders dataset----


,Buy Ts,Revenue,Uid
943,2017-06-08 11:20:00,0.92,10871236662131028680
3571,2017-07-17 10:16:00,10.39,3146056445548451053
43813,2018-04-11 20:10:00,2.08,3272168776000204419
49834,2018-05-31 09:38:00,4.03,2744482237082730837
21993,2017-12-05 10:22:00,1.83,386864823188327628


-------------------------------------------------------------------
----Visits dataset----


,Device,End Ts,Source Id,Start Ts,Uid
39595,touch,2018-05-17 11:11:00,4,2018-05-17 10:45:00,7397240624762011590
22807,desktop,2018-03-22 10:47:00,4,2018-03-22 10:41:00,9166692859537744589
240395,desktop,2017-11-23 19:15:00,5,2017-11-23 19:07:00,16720047451638905628
220728,touch,2017-10-22 00:39:00,4,2017-10-22 00:35:00,14102030181122616087
147436,touch,2018-05-29 11:47:00,2,2018-05-29 11:41:00,11521628752606790669


### Costs dataset

In [4]:
print(f'Row count: {cost.shape[0]}')
print(f'Column count: {cost.shape[1]}')

Row count: 2542
Column count: 3


**This shows that `cost` dataset has 2542 rows and 3 columns**

In [5]:
cost.info()

<class 'pandas.DataFrame'>
RangeIndex: 2542 entries, 0 to 2541
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   source_id  2542 non-null   int64  
 1   dt         2542 non-null   str    
 2   costs      2542 non-null   float64
dtypes: float64(1), int64(1), str(1)
memory usage: 84.5 KB


In [6]:
cost.describe().T

,count,mean,std,min,25%,50%,75%,max
source_id,2542.0,4.857199,3.181581,1.00,2.000,4.000,9.000,10.00
costs,2542.0,129.477427,156.296628,0.54,21.945,77.295,170.065,1788.28


In [7]:
cost.columns.tolist()

['source_id', 'dt', 'costs']

In [8]:
cost.nunique().sort_values()

source_id       7
dt            364
costs        2396
dtype: int64

**This shows that in `Cost` dataset. in each column thw we here have unique values**

In [9]:
cost.isna().sum().sort_values(ascending=False)

source_id    0
dt           0
costs        0
dtype: int64

In [10]:
missing_cost = (cost.isna().mean().mul(100).sort_values(ascending=False))

missing_cost

source_id    0.0
dt           0.0
costs        0.0
dtype: float64

**From this above there is no null values in `costs` dataset**

In [11]:
dup_cost = cost.duplicated().sum()

print(f'Total duplicated rows : {dup_cost}')

Total duplicated rows : 0


### Orders dataset

In [12]:
print(f'Row count: {orders.shape[0]}')
print(f'Column count: {orders.shape[1]}')

Row count: 50415
Column count: 3


From above we clearly say that, in `Order` dataset 50415 rows and 3 columns take place

In [13]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 50415 entries, 0 to 50414
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Buy Ts   50415 non-null  str    
 1   Revenue  50415 non-null  float64
 2   Uid      50415 non-null  uint64 
dtypes: float64(1), str(1), uint64(1)
memory usage: 2.1 MB


In [14]:
orders.describe().T

,count,mean,std,min,25%,50%,75%,max
Revenue,50415.0,4.999647e+00,2.181836e+01,0.000000e+00,1.220000e+00,2.500000e+00,4.890000e+00,2.633280e+03
Uid,50415.0,9.098161e+18,5.285742e+18,3.135781e+14,4.533567e+18,9.102274e+18,1.368290e+19,1.844617e+19


In [15]:
orders.columns.tolist()

['Buy Ts', 'Revenue', 'Uid']

In [16]:
orders.nunique().sort_values()

Revenue     1149
Uid        36523
Buy Ts     45991
dtype: int64

In `Orders` dataset count of unique values are as like above for each column

In [17]:
orders.isna().sum().sort_values(ascending=False)

Buy Ts     0
Revenue    0
Uid        0
dtype: int64

and from this cell it shows us that there is no null cell in `orders` dataset

In [18]:
missing_orders = orders.isna().mean().mul(100).sort_values(ascending=False)
missing_orders

Buy Ts     0.0
Revenue    0.0
Uid        0.0
dtype: float64

In [19]:
dup_orders = orders.duplicated().sum()

print(f'Total duplicated rows : {dup_orders}')

Total duplicated rows : 0


In the `Orders` dataset there is no fully duplicated rows

### Visits dataset

In [20]:
print(f'Row count: {visit.shape[0]}')
print(f'Columns count: {visit.shape[1]}')

Row count: 359400
Columns count: 5


**In the `visits` dataset There are 359400 rows and 5 columns**

In [21]:
visit.info()

<class 'pandas.DataFrame'>
RangeIndex: 359400 entries, 0 to 359399
Data columns (total 5 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   Device     359400 non-null  str   
 1   End Ts     359400 non-null  str   
 2   Source Id  359400 non-null  int64 
 3   Start Ts   359400 non-null  str   
 4   Uid        359400 non-null  uint64
dtypes: int64(1), str(3), uint64(1)
memory usage: 28.9 MB


In [22]:
visit.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Device,359400,2,desktop,262567,NaN,NaN,NaN,NaN,NaN,NaN,NaN
End Ts,359400,224760,2017-11-24 16:51:00,23,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Source Id,359400.0,NaN,NaN,NaN,3.750515,1.917116,1.0,3.0,4.0,5.0,10.0
Start Ts,359400,224303,2017-11-24 16:06:00,19,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Uid,359400.0,NaN,NaN,NaN,9202556994354921472.0,5298432824659611648.0,11863502262781.0,4613406732167550976.0,9227412908740601856.0,13728237011236173824.0,18446676030785673216.0


In [23]:
visit.columns.tolist()

['Device', 'End Ts', 'Source Id', 'Start Ts', 'Uid']

From above we can understad that, there 2 columns (`source_id`,`Uid`) are the non object columns

In [24]:
visit.nunique().sort_values()

Device            2
Source Id         9
Start Ts     224303
End Ts       224760
Uid          228169
dtype: int64

in the last cell 
- `Device` have **2**
- `source id`  have **9**
- `Start ts` have **224303**
- `End ts` have **224760**
- `Uid` have **228169**

unique values

we can say that looking at `source id` and `uid` customers average user visits the site more than 1


In [25]:
visit.isna().sum().sort_values(ascending=False)

Device       0
End Ts       0
Source Id    0
Start Ts     0
Uid          0
dtype: int64

In [26]:
missing_visit = visit.isna().mean().mul(100).sort_values(ascending=False)
missing_visit

Device       0.0
End Ts       0.0
Source Id    0.0
Start Ts     0.0
Uid          0.0
dtype: float64

we can say that there is no null value

In [27]:
dup_visit = visit.duplicated().sum()

print(f'Total duplicated rows: {dup_visit}')

Total duplicated rows: 0


in the `visit` dataset there is no totally duplicated rows

## Data preparation

In [28]:
visit.columns.tolist(), orders.columns.tolist(), cost.columns.tolist()

(['Device', 'End Ts', 'Source Id', 'Start Ts', 'Uid'],
 ['Buy Ts', 'Revenue', 'Uid'],
 ['source_id', 'dt', 'costs'])

In [29]:
visit.columns = ['device','end_ts','source_id','start_ts','uid']
orders.columns = ['buy_ts','revenue','uid']
cost.columns = ['source_id','dt','costs']

In [30]:
visit.columns.tolist(), orders.columns.tolist(), cost.columns.tolist()

(['device', 'end_ts', 'source_id', 'start_ts', 'uid'],
 ['buy_ts', 'revenue', 'uid'],
 ['source_id', 'dt', 'costs'])

in here i corrected column names

In [31]:
print('---Visit dataset---')
print(visit.dtypes)
print()
print('---Orders dataset---')
print(orders.dtypes)
print()
print('---Costs dataset---')
print(cost.dtypes)

---Visit dataset---
device          str
end_ts          str
source_id     int64
start_ts        str
uid          uint64
dtype: object

---Orders dataset---
buy_ts         str
revenue    float64
uid         uint64
dtype: object

---Costs dataset---
source_id      int64
dt               str
costs        float64
dtype: object


in here we can see that some datetime columns are not the correct format

- in `visit` dataset 2 columns(`start_ts`,`end_ts`)
- in `orders` dataset 1 column (`buy_ts`)
- in `cost` dataset 1 column (`dt`)

In [32]:
visit['start_ts'] = pd.to_datetime(visit['start_ts'])
visit['end_ts'] = pd.to_datetime(visit['end_ts'])

orders['buy_ts'] = pd.to_datetime(orders['buy_ts'])

cost['dt'] = pd.to_datetime(cost['dt'])

In [33]:
print('---Visit dataset---')
print(visit.dtypes)
print()
print('---Orders dataset---')
print(orders.dtypes)
print()
print('---Costs dataset---')
print(cost.dtypes)

---Visit dataset---
device                  str
end_ts       datetime64[us]
source_id             int64
start_ts     datetime64[us]
uid                  uint64
dtype: object

---Orders dataset---
buy_ts     datetime64[us]
revenue           float64
uid                uint64
dtype: object

---Costs dataset---
source_id             int64
dt           datetime64[us]
costs               float64
dtype: object


In [34]:
print(f"Visits date range: {visit['start_ts'].min()} - {visit['start_ts'].max()}")
print(f"Orders date range: {orders['buy_ts'].min()} - {orders['buy_ts'].max()}")
print(f"Costs date range: {cost['dt'].min()} - {cost['dt'].max()}")

Visits date range: 2017-06-01 00:01:00 - 2018-05-31 23:59:00
Orders date range: 2017-06-01 00:10:00 - 2018-06-01 00:02:00
Costs date range: 2017-06-01 00:00:00 - 2018-05-31 00:00:00


we corrected the column formats

In [35]:
session = (visit['end_ts']-visit['start_ts']).dt.total_seconds()
print(f'Count of negative session: {(session < 0).sum()}')
visit[session < 0]

Count of negative session: 2


,device,end_ts,source_id,start_ts,uid
4181,desktop,2018-03-25 03:18:00,3,2018-03-25 03:50:00,13092152539246794986
177972,desktop,2018-03-25 03:09:00,9,2018-03-25 03:55:00,4621202742905035453


In these two rows, the end time is earlier than the start time, which is why these two sessions have negative durations.

These two values are also significantly smaller than the total, so I will not drop these rows because they will not have a meaningful impact on the overall results.

Additionally, the other columns in these two rows are correct.


In [36]:
visit['session_duration_sec'] = (visit['end_ts']-visit['start_ts']).dt.total_seconds()

visit.loc[visit['session_duration_sec']<0,'session_duration_sec'] = 0

print(f'Checikng wether there is any negative session : {(visit['session_duration_sec']<0).any()}')

Checikng wether there is any negative session : False


In [37]:
first_activity = visit.groupby('uid').agg(
    first_activity_ts =('start_ts','min'),
    first_device = ('device','first'),
    first_source_id = ('source_id','first')
).reset_index()

first_activity

,uid,first_activity_ts,first_device,first_source_id
0,11863502262781,2018-03-01 17:27:00,touch,3
1,49537067089222,2018-02-06 15:55:00,touch,2
2,297729379853735,2017-06-07 18:47:00,desktop,3
3,313578113262317,2017-09-18 22:49:00,desktop,2
4,325320750514679,2017-09-30 14:29:00,desktop,5
...,...,...,...,...
228164,18446403737806311543,2017-11-30 03:36:00,desktop,5
228165,18446424184725333426,2017-12-06 20:32:00,touch,4
228166,18446556406699109058,2018-01-01 16:29:00,touch,3
228167,18446621818809592527,2017-12-27 13:27:00,desktop,4


In [38]:
first_activity['first_activity_month'] = first_activity['first_activity_ts'].dt.to_period('M').apply(lambda r: r.start_time)

first_activity['first_activity_week'] = first_activity['first_activity_ts'].dt.to_period('W').apply(lambda r: r.start_time)

In [39]:
first_activity.head(5)

,uid,first_activity_ts,first_device,first_source_id,first_activity_month,first_activity_week
0,11863502262781,2018-03-01 17:27:00,touch,3,2018-03-01,2018-02-26
1,49537067089222,2018-02-06 15:55:00,touch,2,2018-02-01,2018-02-05
2,297729379853735,2017-06-07 18:47:00,desktop,3,2017-06-01,2017-06-05
3,313578113262317,2017-09-18 22:49:00,desktop,2,2017-09-01,2017-09-18
4,325320750514679,2017-09-30 14:29:00,desktop,5,2017-09-01,2017-09-25


In [40]:
print(f'Number of unique users (visits): {first_activity['uid'].nunique():,} ')

Number of unique users (visits): 228,169 


in the cells above, i just wanted to see the first visits of the customers in the dataset


## Metrics

### Product metrics

#### DAU WAU MAU

- `DAU` - Dayly active user
- `WAU` - Weeekly active user
- `MAU` - Montly active user

In [41]:
visit['session_date'] = visit['start_ts'].dt.date
visit['session_week'] = visit['start_ts'].dt.to_period('W')
visit['session_month'] = visit['start_ts'].dt.to_period('M')


In [42]:
dau = visit.groupby('session_date')['uid'].nunique().mean()
wau = visit.groupby('session_week')['uid'].nunique().mean()
mau = visit.groupby('session_month')['uid'].nunique().mean()


sticky_factor = dau/mau * 100

In [43]:
print(f'Average DAU : {dau:.0f}')
print(f'Average WAU : {wau:.0f}')
print(f'Average MAU : {mau:.0f}')
print(f'Sticky factor (DAU/MAU): {sticky_factor:.1f}%')

Average DAU : 908
Average WAU : 5716
Average MAU : 23228
Sticky factor (DAU/MAU): 3.9%


In [44]:
dau_daily = visit.groupby('session_date')['uid'].nunique().reset_index()
dau_daily.columns = ['date', 'dau']

In [45]:
fig_daily = px.line(
    dau_daily,
    x='date',
    y='dau',
    title='Number of Daily Active Users',
    labels={
        'date': 'Date',
        'dau': 'Number of Unique Visitors'
    }
)

fig_daily.show()

**Result:** The average sticky factor (DAU/MAU) is approximately **4%**, meaning only 4% of monthly active users visit the site on a typical day. This indicates low daily user retention. The metric rises noticeably toward December, which may suggest **seasonal demand** related to the holiday period [not directly verified].


#### Daily sessions

In [46]:
session_per_day = visit.groupby('session_date')['uid'].count()

print(f'Daily average session count: {session_per_day.mean():.0f}')
print(f'Daily Median session count: {session_per_day.median():.0f}')

Daily average session count: 987
Daily Median session count: 1003


**as a result**

**Result:** The site averages **987 sessions per day** and **908 DAU**, meaning approximately **1.09 sessions per user per day**. This suggests that users rarely return multiple times within the same day.


#### Average Session lengt

In [47]:
lengt = visit['session_duration_sec'].describe()
length_mode = visit['session_duration_sec'].mode()[0]

print(lengt)
print(f'The frequancy of session: {length_mode:.0f}')

count    359400.000000
mean        643.038709
std         997.103591
min           0.000000
25%         120.000000
50%         300.000000
75%         840.000000
max       42660.000000
Name: session_duration_sec, dtype: float64
The frequancy of session: 60


In [48]:
fig = px.histogram(
    visit[visit['session_duration_sec'] < 1800], x='session_duration_sec',
    nbins=60,
    title='Session duration distribution',
    labels={'session_duration_sec': 'Session_timer', 'count': 'Sessiya sayı'}
)
fig.show()

**Result:** The average session lasts **644 seconds (10.7 minutes)**, while the median is only **300 seconds (5 minutes)**. This indicates a **right-skewed distribution**, where a small number of very long sessions increases the average. The most common duration is **60 seconds**, suggesting that many users visit the site only briefly. Therefore, the **median is a more reliable measure** of typical session duration.


#### Retention rate

In [49]:
first_activity = visit.groupby('uid')['session_month'].min().reset_index()
first_activity.columns = ['uid', 'first_activity_month']

In [50]:
visit_cohort = visit.merge(first_activity, on='uid')
visit_cohort['cohort_age'] = (
    visit_cohort['session_month'] - visit_cohort['first_activity_month']
)

In [51]:

cohort_sizes = visit_cohort.groupby('first_activity_month')['uid'].nunique()
cohort_sizes.name = 'n_users'


In [52]:

visit_cohort = visit.merge(first_activity, on='uid')

visit_cohort['cohort_age'] = (
    (visit_cohort['session_month'].dt.year -
     visit_cohort['first_activity_month'].dt.year) * 12
    + visit_cohort['session_month'].dt.month
    - visit_cohort['first_activity_month'].dt.month
).astype('int64')


In [53]:

cohort_activity = (
    visit_cohort
    .groupby(['first_activity_month', 'cohort_age'])['uid']
    .nunique()
    .reset_index(name='n_active')
)

cohort_activity = cohort_activity.merge(
    cohort_sizes.rename('n_users').reset_index(),
    on='first_activity_month',
    how='left',
    validate='many_to_one'
)

cohort_activity['retention'] = (
    cohort_activity['n_active'] / cohort_activity['n_users']
)
cohort_activity['retention'] = cohort_activity['n_active'] / cohort_activity['n_users']


In [54]:

retention_pivot = cohort_activity.pivot_table(
    index='first_activity_month', columns='cohort_age', values='retention'
)
retention_pivot.round(3)

cohort_age,0,1,2,3,4,5,6,7,8,9,10,11
first_activity_month,,,,,,,,,,,,
2017-06,1.0,0.079,0.054,0.061,0.069,0.071,0.061,0.058,0.052,0.051,0.041,0.045
2017-07,1.0,0.056,0.051,0.056,0.058,0.048,0.045,0.046,0.039,0.029,0.027,NaN
2017-08,1.0,0.077,0.063,0.063,0.050,0.044,0.036,0.039,0.028,0.026,NaN,NaN
2017-09,1.0,0.085,0.069,0.051,0.039,0.038,0.036,0.024,0.023,NaN,NaN,NaN
2017-10,1.0,0.079,0.052,0.039,0.034,0.032,0.021,0.020,NaN,NaN,NaN,NaN
2017-11,1.0,0.078,0.044,0.039,0.034,0.023,0.022,NaN,NaN,NaN,NaN,NaN
2017-12,1.0,0.056,0.038,0.031,0.020,0.019,NaN,NaN,NaN,NaN,NaN,NaN
2018-01,1.0,0.060,0.039,0.025,0.020,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2018-02,1.0,0.057,0.025,0.020,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [55]:
retention_pivot_copy = retention_pivot.copy()
retention_pivot_copy.index = retention_pivot_copy.index.astype(str)

In [56]:
retention_fig = px.imshow(
    retention_pivot_copy,
    text_auto='.0%',
    aspect='auto',
    color_continuous_scale='Blues',
    title='User Retention Matrix — Month × Cohort',
    labels={
        'x': 'Cohort ',
        'y': 'The first activity month (cohort)',
        'color': 'Retention'
    }
)

retention_fig.show()

Retention drops sharply to **~7% in Month 2** and stabilizes at **~3–5%** thereafter, indicating **low long-term retention** and predominantly event-specific, one-time usage.


### Sales metrics

#### Conversion Time

In [57]:
first_visit = visit.groupby('uid')['start_ts'].min().reset_index()
first_visit.columns = ['uid','first_visit_ts']
first_visit.head(5)

,uid,first_visit_ts
0,11863502262781,2018-03-01 17:27:00
1,49537067089222,2018-02-06 15:55:00
2,297729379853735,2017-06-07 18:47:00
3,313578113262317,2017-09-18 22:49:00
4,325320750514679,2017-09-30 14:29:00


In [58]:
first_order = orders.groupby('uid')['buy_ts'].min().reset_index()
first_order.columns = ['uid', 'first_order_ts']
first_order.head(5)

,uid,first_order_ts
0,313578113262317,2018-01-03 21:51:00
1,1575281904278712,2017-06-03 10:13:00
2,2429014661409475,2017-10-11 18:33:00
3,2464366381792757,2018-01-28 15:54:00
4,2551852515556206,2017-11-24 10:14:00


In [59]:
conversion = first_order.merge(first_visit,on='uid',how='left')
conversion.head(5)

,uid,first_order_ts,first_visit_ts
0,313578113262317,2018-01-03 21:51:00,2017-09-18 22:49:00
1,1575281904278712,2017-06-03 10:13:00,2017-06-03 10:13:00
2,2429014661409475,2017-10-11 18:33:00,2017-10-11 17:14:00
3,2464366381792757,2018-01-28 15:54:00,2018-01-27 20:10:00
4,2551852515556206,2017-11-24 10:14:00,2017-11-24 10:14:00


In [60]:
conversion['conversion_days'] = (conversion['first_order_ts']-conversion['first_visit_ts']).dt.days

conversion.sample(5)

,uid,first_order_ts,first_visit_ts,conversion_days
29290,14775556297012766849,2017-10-31 21:02:00,2017-10-31 20:54:00,0
23665,11951104727477909149,2018-02-07 11:37:00,2018-02-07 11:24:00,0
32715,16517298957294701980,2017-10-09 19:00:00,2017-10-09 18:58:00,0
9099,4596952703214662934,2017-08-14 08:28:00,2017-08-14 08:25:00,0
5716,2889993679184661255,2018-05-18 11:37:00,2018-05-18 10:04:00,0


In [61]:
print(f"Number of buyers with no first visit found: {conversion['first_visit_ts'].isna().sum()}")
conversion['conversion_days'].describe().round(2)

Number of buyers with no first visit found: 0


count    36523.00
mean        16.73
std         46.96
min          0.00
25%          0.00
50%          0.00
75%          2.00
max        363.00
Name: conversion_days, dtype: float64

In [62]:
fig = px.histogram(
    conversion[conversion['conversion_days'] <= 30], x='conversion_days',
    nbins=31,
    title='Distribution of Days from First Visit to First Purchase (0–30 Days)',
    labels={'conversion_days': 'Conversion Days', 'count': 'Number of Users'}
)
fig.show()

**Result:** **50% of buyers** purchase on the same day as their first visit, indicating strong **impulse-buying behavior**. The **16.7-day mean** is skewed upward by late purchases, making the **median more representative**.


#### Number of orders in a period

In [63]:
orders['order_month'] = orders['buy_ts'].dt.to_period('M')
order_per_month = orders.groupby('order_month')['uid'].count()

In [64]:
num_of_buyers = orders['uid'].nunique()
num_of_orders = len(orders)

In [65]:
print(f'Total unique buyer count: {num_of_buyers}')
print(f'Total orders count: {num_of_orders}')
print(f'Average number of orders per buyer: {num_of_orders/num_of_buyers:.2f}')

Total unique buyer count: 36523
Total orders count: 50415
Average number of orders per buyer: 1.38


In [66]:
fig_order_per_m = px.bar(
    x= order_per_month.index.astype(str), 
    y= order_per_month.values,
    title= 'Monthly sales',
    labels= {'x': 'Month','y':'Number of orders'}
)

fig_order_per_m.show()

**Result**: Buyers place an average of 1.38 orders, indicating low repeat purchasing, with orders peaking toward December in line with the DAU trend.

#### LTV (Customer Lifetime Value)

In [67]:
orders_ltv = orders.merge(first_visit,on='uid',how='left')
orders_ltv.sample(5)

,buy_ts,revenue,uid,order_month,first_visit_ts
28088,2018-01-04 19:19:00,6.35,2372603640946287896,2018-01,2017-12-29 12:40:00
24983,2017-12-19 23:55:00,7.94,14357016286870891288,2017-12,2017-12-17 23:46:00
42129,2018-03-26 19:29:00,3.05,12794745207747551731,2018-03,2018-03-26 19:29:00
6388,2017-08-30 15:54:00,1.17,3824054258858370375,2017-08,2017-08-30 15:50:00
34475,2018-02-13 08:20:00,9.78,4376434777169552154,2018-02,2018-02-13 07:37:00


In [68]:
orders_ltv['cohort_month'] = orders_ltv['first_visit_ts'].dt.to_period('M')
orders_ltv['month_number'] = (orders_ltv['order_month']  - orders_ltv['cohort_month']).apply(lambda x: x.n)

In [69]:
orders_ltv.sample(5)

,buy_ts,revenue,uid,order_month,first_visit_ts,cohort_month,month_number
36028,2018-02-20 21:02:00,4.28,4477532852572828781,2018-02,2018-02-18 22:09:00,2018-02,0
13408,2017-10-18 14:39:00,3.67,6821654442090722108,2017-10,2017-09-22 15:56:00,2017-09,1
22841,2017-12-08 16:11:00,4.89,2768288399353436468,2017-12,2017-11-02 14:04:00,2017-11,1
3943,2017-07-19 14:52:00,3.05,2632899193834751796,2017-07,2017-07-18 05:57:00,2017-07,0
7476,2017-09-11 15:36:00,0.73,4352012736328198047,2017-09,2017-09-11 15:25:00,2017-09,0


In [70]:
size_of_buyers_cohort = (
    first_visit.assign(cohort_month=first_visit['first_visit_ts'].dt.to_period('M'))
    .groupby('cohort_month')['uid']
    .nunique()
    .rename('n_buyers')
)

size_of_buyers_cohort

cohort_month
2017-06    13259
2017-07    13140
2017-08    10181
2017-09    16704
2017-10    25977
2017-11    27248
2017-12    25268
2018-01    22624
2018-02    22197
2018-03    20589
2018-04    15709
2018-05    15273
Freq: M, Name: n_buyers, dtype: int64

In [71]:
revenue_by_month_num = orders_ltv.groupby(['cohort_month', 'month_number'])['revenue'].sum().reset_index()
revenue_by_month_num.sample(5)

,cohort_month,month_number,revenue
27,2017-08,4,810.04
24,2017-08,1,997.32
10,2017-06,10,2204.01
35,2017-09,2,2110.63
65,2018-01,2,1164.76


In [72]:
revenue_by_month_num = revenue_by_month_num.merge(size_of_buyers_cohort,on='cohort_month')
revenue_by_month_num.sample(5)

,cohort_month,month_number,revenue,n_buyers
18,2017-07,6,588.00,13140
7,2017-06,7,2270.81,13259
64,2018-01,1,1951.70,22624
61,2017-12,4,1457.57,25268
13,2017-07,1,1059.28,13140


In [73]:
revenue_by_month_num['lifetime_value'] = (revenue_by_month_num['revenue']/revenue_by_month_num['n_buyers']).round(5)

revenue_by_month_num.sample(5)

,cohort_month,month_number,revenue,n_buyers,lifetime_value
61,2017-12,4,1457.57,25268,0.05768
45,2017-10,3,1058.05,25977,0.04073
77,2018-05,0,10368.11,15273,0.67885
4,2017-06,4,3940.59,13259,0.29720
11,2017-06,11,1927.81,13259,0.14540


In [74]:
ltv_pivot = revenue_by_month_num.pivot_table(index='cohort_month',columns='month_number',values='lifetime_value')
ltv_cumulative = ltv_pivot.cumsum(axis=1)

print(ltv_cumulative)

month_number       0        1        2        3        4        5        6   \
cohort_month                                                                  
2017-06       0.72083  0.92179  1.03226  1.32505  1.62225  1.83527  2.09848   
2017-07       0.75151  0.83212  0.91651  0.99132  1.05578  1.11652  1.16127   
2017-08       0.61240  0.71036  0.80871  0.90165  0.98121  1.03355  1.08135   
2017-09       0.73977  0.96130  1.08765  1.71259  1.78973  1.90408  1.96196   
2017-10       0.70687  0.85353  0.90461  0.94534  0.98551  1.02486  1.04715   
2017-11       0.60672  0.70462  0.75390  0.80416  0.84485  0.87082  0.89646   
2017-12       0.66713  0.73595  0.90210  1.10488  1.16256  1.23309      NaN   
2018-01       0.46840  0.55467  0.60615  0.61967  0.64298      NaN      NaN   
2018-02       0.51866  0.60806  0.63794  0.67853      NaN      NaN      NaN   
2018-03       0.58339  0.63118  0.66881      NaN      NaN      NaN      NaN   
2018-04       0.51205  0.62910      NaN      NaN    

In [75]:
ltv_heatmap = ltv_cumulative.copy().fillna(0)

ltv_heatmap.index = ltv_heatmap.index.astype(str)
ltv_heatmap.columns = [str(col) for col in ltv_heatmap.columns]

fig_heatmap = px.imshow(
    ltv_heatmap,
    text_auto=".2f",
    aspect="auto",
    color_continuous_scale="Blues",
    title="Cumulative LTV by Cohort and Month",
    labels={
        "x": "Month number",
        "y": "Cohort month",
        "color": "Cumulative LTV"
    }
)

fig_heatmap.update_layout(
    xaxis_title="Month number",
    yaxis_title="Cohort month",
    coloraxis_colorbar=dict(title="Cumulative LTV")
)

fig_heatmap.show()

### Marketing metrics

#### Marketing costs

In [76]:
marketing_cost = cost['costs'].sum()

In [77]:
cost_by_source_id = cost.groupby('source_id')['costs'].sum().sort_values(ascending= False)

cost_by_source_id.head(5)

source_id
3    141321.63
4     61073.60
5     51757.10
2     42806.04
1     20833.27
Name: costs, dtype: float64

In [78]:
cost['cost_month'] = cost['dt'].dt.to_period('M')
cost_by_month = cost.groupby('cost_month')['costs'].sum()
cost_by_month.head(5)

cost_month
2017-06    18015.00
2017-07    18240.59
2017-08    14790.54
2017-09    24368.91
2017-10    36322.88
Freq: M, Name: costs, dtype: float64

In [79]:
print(f'Total marketing costs: {marketing_cost:.2f}')
print('-----------------------------------------')
print('Cost by source')
print(cost_by_source_id)

Total marketing costs: 329131.62
-----------------------------------------
Cost by source
source_id
3     141321.63
4      61073.60
5      51757.10
2      42806.04
1      20833.27
10      5822.49
9       5517.49
Name: costs, dtype: float64


In [82]:
fig_costs = px.bar(
    x=cost_by_source_id.index.astype(str),
    y=cost_by_source_id.values,
    title='Marketing costs by source id',
    labels={'x': 'Source id', 'y': 'Costs'}
)

fig_costs.show()

In [83]:
fig = px.line(
    x=cost_by_month.index.astype(str), y=cost_by_month.values,
    title='Costs by months',
    labels={'x': 'Month', 'y': 'Cost'},
    markers=True
)
fig.show()

**Result:** Total marketing spend is **$329K**, with **Source 3** receiving the highest investment at **$141K**; spending also rises toward **December**, aligning with the DAU and order trends.


#### CAC (Customer Acquisition Cost)

In [84]:
buyer_source = visit[['uid', 'source_id']].drop_duplicates(subset='uid')
buyers_with_source = first_order.merge(buyer_source, on='uid', how='left')

In [87]:
n_buyers_by_source = buyers_with_source.groupby('source_id')['uid'].nunique()
cac_by_source = (cost_by_source_id / n_buyers_by_source).dropna().sort_values()


cac_by_source

source_id
9      3.899286
10     4.691773
1      5.565928
4      6.217408
5      7.567934
2      9.901929
3     15.465269
dtype: float64

In [88]:
fig = px.bar(
    x=cac_by_source.index.astype(str), y=cac_by_source.values,
    title='Customer Acquisition Cost',
    labels={'x': 'Source ID', 'y': 'CAC'}
)
fig.show()

Result: CAC varies nearly 4× across sources, from ~$3.9 (Source 9) to ~$15.5 (Source 3); Source 3 has both the highest spend and highest CAC, raising concerns about budget efficiency.

#### ROI (Return On Investment)

In [90]:
cost.columns.to_list()

['source_id', 'dt', 'costs', 'cost_month']

In [92]:
cost_by_cohort_month = cost.groupby('cost_month')['costs'].sum()
cac_by_cohort_month = cost_by_cohort_month/cohort_sizes

cac_by_cohort_month.head(5)

cost_month
2017-06    1.358700
2017-07    1.388173
2017-08    1.452759
2017-09    1.458867
2017-10    1.398271
Freq: M, dtype: float64

In [94]:
roi = (ltv_cumulative.divide(cac_by_cohort_month,axis=0)).round(5)

roi

month_number,0,1,2,3,4,5,6,7,8,9,10,11
cohort_month,,,,,,,,,,,,
2017-06,0.53053,0.67844,0.75974,0.97523,1.19397,1.35075,1.54448,1.67053,1.79920,1.99246,2.11481,2.22182
2017-07,0.54137,0.59944,0.66023,0.71412,0.76055,0.80431,0.83655,0.87877,0.91850,0.95090,0.97651,NaN
2017-08,0.42154,0.48897,0.55667,0.62065,0.67541,0.71144,0.74434,0.82860,0.88476,0.91016,NaN,NaN
2017-09,0.50709,0.65894,0.74554,1.17392,1.22679,1.30518,1.34485,1.36502,1.38220,NaN,NaN,NaN
2017-10,0.50553,0.61042,0.64695,0.67608,0.70481,0.73295,0.74889,0.76695,NaN,NaN,NaN,NaN
2017-11,0.43611,0.50648,0.54190,0.57803,0.60727,0.62594,0.64437,NaN,NaN,NaN,NaN,NaN
2017-12,0.43996,0.48534,0.59491,0.72864,0.76668,0.81319,NaN,NaN,NaN,NaN,NaN,NaN
2018-01,0.31616,0.37439,0.40913,0.41826,0.43399,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2018-02,0.35182,0.41247,0.43273,0.46027,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [95]:
roi_plot_data = roi.iloc[:, :6].copy()
roi_plot_data.index = roi_plot_data.index.astype(str)


In [97]:

fig = px.line(
    roi_plot_data.T,
    title='Cumulative ROI by Cohort — First 6 Months (1.0 = Break-Even)',
    labels={'index': 'Kohortun yaşı (ay)', 'value': 'ROI', 'variable': 'Kohort (ilk vizit ayı)'}
)
fig.add_hline(y=1.0, line_dash='dash', line_color='red',
              annotation_text='Özünü doğrultma həddi (ROI=1.0)')
fig.update_xaxes(title='Kohortun yaşı (ay)')
fig.update_yaxes(title='ROI')
fig.show()

**Result:** No cohort reaches **ROI = 1.0 within 6 months**. The best-performing cohorts reach **~1.3–1.35**, while the weakest remain at **~0.4–0.45**, indicating substantial variation in marketing efficiency across cohorts.


### Devices

In [98]:
device_summary = visit.groupby('device').agg(
    n_sessions=('uid', 'count'),
    avg_session_sec=('session_duration_sec', 'mean')
).reset_index()

In [99]:
device_orders = orders.merge(
    visit[['uid', 'device']].drop_duplicates(subset='uid'), on='uid', how='left'
)

In [100]:

device_revenue = device_orders.groupby('device')['revenue'].agg(['mean', 'sum']).reset_index()
device_revenue.columns = ['device', 'avg_check', 'total_revenue']

device_summary = device_summary.merge(device_revenue, on='device')
device_summary

,device,n_sessions,avg_session_sec,avg_check,total_revenue
0,desktop,262567,703.306684,5.171421,211531.81
1,touch,96833,479.619407,4.260897,40525.39


In [101]:
fig = px.bar(
    device_summary, x='device', y='n_sessions',
    title='Cihaz tipinə görə sessiya sayı',
    labels={'device': 'Cihaz', 'n_sessions': 'Sessiya sayı'}
)
fig.show()

**Result:** Desktop accounts for the majority of sessions, indicating greater user activity on desktop than touch devices. However, further A/B testing is needed before prioritizing desktop-focused marketing and landing page optimization.


## Additional

In [106]:
daily_costs = cost.groupby('dt')['costs'].sum()
daily_revenue = orders.groupby(orders['buy_ts'].dt.date)['revenue'].sum()
daily_revenue.index = pd.to_datetime(daily_revenue.index)

In [107]:

merged_daily = pd.concat([daily_costs.rename('cost'), daily_revenue.rename('revenue')], axis=1).dropna()

corr, p_val = st.spearmanr(merged_daily['cost'], merged_daily['revenue'])
print(f"Spearmen correlation value: {corr:.3f} (p-value={p_val:.4f})")

Spearmen correlation value: 0.721 (p-value=0.0000)


**Result:** Positive correlation between spend and revenue does not prove causation, as both may be driven by seasonal demand. **Controlled experiments** are needed to confirm the impact of marketing spend.
